fpga_interchange/site_arch.h

```diff
@@ -0,0 +1,990 @@
+    PortType getBelPinType(BelId bel, IdString pin) const NPNR_ATTRIBUTE(__always_inline__);
+
+    const char *nameOfWire(const SiteWire &wire) const;
+    const char *nameOfPip(const SitePip &pip) const;
+    const char *nameOfNet(const SiteNetInfo *net) const;
+
+    bool debug() const;
+
+    bool bindWire(const SiteWire &wire, SiteNetInfo *net)
+    {
+        auto result = wire_to_nets.emplace(wire, SiteNetMap{net, 1});
+        if (result.first->second.net != net) {
+            if (debug()) {
+                log_info("Net conflict binding wire %s to net %s, conflicts with net %s\n", nameOfWire(wire),
+                         nameOfNet(net), nameOfNet(result.first->second.net));
+            }
+            return false;
+        }
+
+        if (!result.second) {
+            result.first->second.count += 1;
+        }
+
+        return true;
+    }
+
+    SiteNetInfo *unbindWire(const SiteWire &wire)
+    {
+        auto iter = wire_to_nets.find(wire);
+        NPNR_ASSERT(iter != wire_to_nets.end());
+        NPNR_ASSERT(iter->second.count >= 1);
+        SiteNetInfo *net = iter->second.net;
+        iter->second.count -= 1;
+
+        if (iter->second.count == 0) {
+            wire_to_nets.erase(iter);
+        }
+
+        return net;
+    }
+
+    bool bindPip(const SitePip &pip, SiteNetInfo *net)
+    {
+        SiteWire src = getPipSrcWire(pip);
+        SiteWire dst = getPipDstWire(pip);
+
+        if (!bindWire(src, net)) {
+            return false;
+        }
+        if (!bindWire(dst, net)) {
+            unbindWire(src);
+            return false;
+        }
+
+        auto result = net->wires.emplace(dst, SitePipMap{pip, 1});
+        if (!result.second) {
+            if (result.first->second.pip != pip) {
+                // Pip conflict!
+                if (debug()) {
+                    log_info("Pip conflict binding pip %s to wire %s, conflicts with pip %s\n", nameOfPip(pip),
+                             nameOfWire(dst), nameOfPip(result.first->second.pip));
+                }
+
+                unbindWire(src);
+                unbindWire(dst);
+                return false;
+            }
+
+            result.first->second.count += 1;
+        }
+
+        return true;
+    }
+
+    void unbindPip(const SitePip &pip)
+    {
+        SiteWire src = getPipSrcWire(pip);
+        SiteWire dst = getPipDstWire(pip);
+
+        SiteNetInfo *src_net = unbindWire(src);
+        SiteNetInfo *dst_net = unbindWire(dst);
+        NPNR_ASSERT(src_net == dst_net);
+        auto iter = dst_net->wires.find(dst);
+        NPNR_ASSERT(iter != dst_net->wires.end());
+        NPNR_ASSERT(iter->second.count >= 1);
+        iter->second.count -= 1;
+
+        if (iter->second.count == 0) {
+            dst_net->wires.erase(iter);
+        }
+    }
+
+    void archcheck();
+
+    bool is_pip_synthetic(const SitePip &pip) const NPNR_ATTRIBUTE(__always_inline__);
+};
+
+struct SitePipDownhillIterator
```

Comment:
It seems at a glance that there's quite a bit of duplication between these iterators that could be tidied up?

Reply:
So `SitePipDownhillIterator` is basically the hottest part of the site router expansion loop, and it is very sensitivity.  I'm not sure how to both reduce the duplication, but also not increase the runtime.  For better or worse, I think `SitePipDownhillIterator` needs to be seperate from the other two iterators because of how critical path it ends up being to the site router.

`PipDownhillIterator` is in a similiar boat, but I haven't taken the time to optimize it as much.